{"cells":[{"cell_type":"code","execution_count":2,"metadata":{"id":"LetKTseWB7iS"},"outputs":[{"name":"stdout","output_type":"stream","text":["\u001B[33mWARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv\u001B[0m\u001B[33m\n","\u001B[0m\u001B[33mWARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv\u001B[0m\u001B[33m\n","\u001B[0m"]}],"source":["!pip install -q google-cloud-storage==1.43.0\n","!pip install -q graphframes"]},{"cell_type":"code","execution_count":3,"metadata":{},"outputs":[],"source":["from graphframes import *\n","from google.cloud import storage\n","import pickle\n","import pyspark"]},{"cell_type":"code","execution_count":4,"metadata":{"id":"M-jcRdy-qL8P"},"outputs":[],"source":["# Put your bucket name below and make sure you can access it without an error\n","bucket_name = 'project_ir313254724_322769175_pagerank'\n","full_path = f\"gs://{bucket_name}/\"\n","paths=[]\n","\n","client = storage.Client()\n","blobs = client.list_blobs(bucket_name)\n","for b in blobs:\n","    if b.name != 'graphframes.sh':\n","        paths.append(full_path+b.name)"]},{"cell_type":"code","execution_count":5,"metadata":{"id":"EbCJ4_hFCLSa"},"outputs":[{"name":"stderr","output_type":"stream","text":["                                                                                \r"]}],"source":["parquetFile = spark.read.parquet(*paths)\n","pages_links = parquetFile.select(\"id\", \"anchor_text\").rdd"]},{"cell_type":"code","execution_count":6,"metadata":{"id":"vkyqsV5nCZqH"},"outputs":[],"source":["def generate_graph(pages):\n","  ''' Compute the directed graph generated by wiki links.\n","  Parameters:\n","  -----------\n","    pages: RDD\n","      An RDD where each row consists of one wikipedia articles with 'id' and\n","      'anchor_text'.\n","  Returns:\n","  --------\n","    edges: RDD\n","      An RDD where each row represents an edge in the directed graph created by\n","      the wikipedia links. The first entry should the source page id and the\n","      second entry is the destination page id. No duplicates should be present.\n","    vertices: RDD\n","      An RDD where each row represents a vetrix (node) in the directed graph\n","      created by the wikipedia links. No duplicates should be present.\n","  '''\n","  # Extract links from Wikipedia articles and flatten the result\n","  edges_rdd = pages.flatMap(lambda row: ((row.id, anchor.id) for anchor in row.anchor_text))\n","\n","  # Remove duplicates from the edges RDD\n","  edges_rdd = edges_rdd.distinct()\n","\n","  # Create vertices RDD by collecting unique page IDs from edges\n","  vertices_rdd = (edges_rdd\n","                  .flatMap(lambda x: (x[0], x[1]))\n","                  .distinct()\n","                  .map(lambda x: (x,)))\n","\n","  return edges_rdd, vertices_rdd"]},{"cell_type":"code","execution_count":7,"metadata":{"id":"uxUTm7KOCaQ6"},"outputs":[{"name":"stderr","output_type":"stream","text":["24/03/08 05:55:39 WARN YarnAllocator: Container from a bad node: container_1709834798656_0014_01_000007 on host: project-ir-selin-dior-w-3.c.project-ir-413718.internal. Exit status: 137. Diagnostics: [2024-03-08 05:55:39.382]Container killed on request. Exit code is 137\n","[2024-03-08 05:55:39.382]Container exited with a non-zero exit code 137. \n","[2024-03-08 05:55:39.383]Killed by external signal\n",".\n","24/03/08 05:55:39 WARN YarnSchedulerBackend$YarnSchedulerEndpoint: Requesting driver to remove executor 6 for reason Container from a bad node: container_1709834798656_0014_01_000007 on host: project-ir-selin-dior-w-3.c.project-ir-413718.internal. Exit status: 137. Diagnostics: [2024-03-08 05:55:39.382]Container killed on request. Exit code is 137\n","[2024-03-08 05:55:39.382]Container exited with a non-zero exit code 137. \n","[2024-03-08 05:55:39.383]Killed by external signal\n",".\n","24/03/08 05:55:39 ERROR YarnScheduler: Lost executor 6 on project-ir-selin-dior-w-3.c.project-ir-413718.internal: Container from a bad node: container_1709834798656_0014_01_000007 on host: project-ir-selin-dior-w-3.c.project-ir-413718.internal. Exit status: 137. Diagnostics: [2024-03-08 05:55:39.382]Container killed on request. Exit code is 137\n","[2024-03-08 05:55:39.382]Container exited with a non-zero exit code 137. \n","[2024-03-08 05:55:39.383]Killed by external signal\n",".\n","24/03/08 05:55:39 WARN TaskSetManager: Lost task 33.0 in stage 2.0 (TID 94) (project-ir-selin-dior-w-3.c.project-ir-413718.internal executor 6): ExecutorLostFailure (executor 6 exited caused by one of the running tasks) Reason: Container from a bad node: container_1709834798656_0014_01_000007 on host: project-ir-selin-dior-w-3.c.project-ir-413718.internal. Exit status: 137. Diagnostics: [2024-03-08 05:55:39.382]Container killed on request. Exit code is 137\n","[2024-03-08 05:55:39.382]Container exited with a non-zero exit code 137. \n","[2024-03-08 05:55:39.383]Killed by external signal\n",".\n","24/03/08 05:56:08 WARN YarnAllocator: Container from a bad node: container_1709834798656_0014_01_000004 on host: project-ir-selin-dior-w-2.c.project-ir-413718.internal. Exit status: 143. Diagnostics: [2024-03-08 05:56:08.220]Container killed on request. Exit code is 143\n","[2024-03-08 05:56:08.220]Container exited with a non-zero exit code 143. \n","[2024-03-08 05:56:08.221]Killed by external signal\n",".\n","24/03/08 05:56:08 WARN YarnSchedulerBackend$YarnSchedulerEndpoint: Requesting driver to remove executor 3 for reason Container from a bad node: container_1709834798656_0014_01_000004 on host: project-ir-selin-dior-w-2.c.project-ir-413718.internal. Exit status: 143. Diagnostics: [2024-03-08 05:56:08.220]Container killed on request. Exit code is 143\n","[2024-03-08 05:56:08.220]Container exited with a non-zero exit code 143. \n","[2024-03-08 05:56:08.221]Killed by external signal\n",".\n","24/03/08 05:56:08 ERROR YarnScheduler: Lost executor 3 on project-ir-selin-dior-w-2.c.project-ir-413718.internal: Container from a bad node: container_1709834798656_0014_01_000004 on host: project-ir-selin-dior-w-2.c.project-ir-413718.internal. Exit status: 143. Diagnostics: [2024-03-08 05:56:08.220]Container killed on request. Exit code is 143\n","[2024-03-08 05:56:08.220]Container exited with a non-zero exit code 143. \n","[2024-03-08 05:56:08.221]Killed by external signal\n",".\n","24/03/08 05:56:08 WARN TaskSetManager: Lost task 36.0 in stage 2.0 (TID 97) (project-ir-selin-dior-w-2.c.project-ir-413718.internal executor 3): ExecutorLostFailure (executor 3 exited caused by one of the running tasks) Reason: Container from a bad node: container_1709834798656_0014_01_000004 on host: project-ir-selin-dior-w-2.c.project-ir-413718.internal. Exit status: 143. Diagnostics: [2024-03-08 05:56:08.220]Container killed on request. Exit code is 143\n","[2024-03-08 05:56:08.220]Container exited with a non-zero exit code 143. \n","[2024-03-08 05:56:08.221]Killed by external signal\n",".\n","24/03/08 06:03:13 WARN YarnAllocator: Container from a bad node: container_1709834798656_0014_01_000006 on host: project-ir-selin-dior-w-3.c.project-ir-413718.internal. Exit status: 143. Diagnostics: [2024-03-08 06:03:13.592]Container killed on request. Exit code is 143\n","[2024-03-08 06:03:13.592]Container exited with a non-zero exit code 143. \n","[2024-03-08 06:03:13.593]Killed by external signal\n",".\n","24/03/08 06:03:13 WARN YarnSchedulerBackend$YarnSchedulerEndpoint: Requesting driver to remove executor 5 for reason Container from a bad node: container_1709834798656_0014_01_000006 on host: project-ir-selin-dior-w-3.c.project-ir-413718.internal. Exit status: 143. Diagnostics: [2024-03-08 06:03:13.592]Container killed on request. Exit code is 143\n","[2024-03-08 06:03:13.592]Container exited with a non-zero exit code 143. \n","[2024-03-08 06:03:13.593]Killed by external signal\n",".\n","24/03/08 06:03:13 ERROR YarnScheduler: Lost executor 5 on project-ir-selin-dior-w-3.c.project-ir-413718.internal: Container from a bad node: container_1709834798656_0014_01_000006 on host: project-ir-selin-dior-w-3.c.project-ir-413718.internal. Exit status: 143. Diagnostics: [2024-03-08 06:03:13.592]Container killed on request. Exit code is 143\n","[2024-03-08 06:03:13.592]Container exited with a non-zero exit code 143. \n","[2024-03-08 06:03:13.593]Killed by external signal\n",".\n","24/03/08 06:03:13 WARN TaskSetManager: Lost task 71.0 in stage 2.0 (TID 134) (project-ir-selin-dior-w-3.c.project-ir-413718.internal executor 5): ExecutorLostFailure (executor 5 exited caused by one of the running tasks) Reason: Container from a bad node: container_1709834798656_0014_01_000006 on host: project-ir-selin-dior-w-3.c.project-ir-413718.internal. Exit status: 143. Diagnostics: [2024-03-08 06:03:13.592]Container killed on request. Exit code is 143\n","[2024-03-08 06:03:13.592]Container exited with a non-zero exit code 143. \n","[2024-03-08 06:03:13.593]Killed by external signal\n",".\n","24/03/08 06:12:04 WARN YarnAllocator: Container from a bad node: container_1709834798656_0014_01_000010 on host: project-ir-selin-dior-w-3.c.project-ir-413718.internal. Exit status: 143. Diagnostics: [2024-03-08 06:12:04.653]Container killed on request. Exit code is 143\n","[2024-03-08 06:12:04.657]Container exited with a non-zero exit code 143. \n","[2024-03-08 06:12:04.657]Killed by external signal\n",".\n","24/03/08 06:12:04 WARN YarnSchedulerBackend$YarnSchedulerEndpoint: Requesting driver to remove executor 9 for reason Container from a bad node: container_1709834798656_0014_01_000010 on host: project-ir-selin-dior-w-3.c.project-ir-413718.internal. Exit status: 143. Diagnostics: [2024-03-08 06:12:04.653]Container killed on request. Exit code is 143\n","[2024-03-08 06:12:04.657]Container exited with a non-zero exit code 143. \n","[2024-03-08 06:12:04.657]Killed by external signal\n",".\n","24/03/08 06:12:04 ERROR YarnScheduler: Lost executor 9 on project-ir-selin-dior-w-3.c.project-ir-413718.internal: Container from a bad node: container_1709834798656_0014_01_000010 on host: project-ir-selin-dior-w-3.c.project-ir-413718.internal. Exit status: 143. Diagnostics: [2024-03-08 06:12:04.653]Container killed on request. Exit code is 143\n","[2024-03-08 06:12:04.657]Container exited with a non-zero exit code 143. \n","[2024-03-08 06:12:04.657]Killed by external signal\n",".\n","24/03/08 06:12:04 WARN TaskSetManager: Lost task 123.0 in stage 3.0 (TID 311) (project-ir-selin-dior-w-3.c.project-ir-413718.internal executor 9): ExecutorLostFailure (executor 9 exited caused by one of the running tasks) Reason: Container from a bad node: container_1709834798656_0014_01_000010 on host: project-ir-selin-dior-w-3.c.project-ir-413718.internal. Exit status: 143. Diagnostics: [2024-03-08 06:12:04.653]Container killed on request. Exit code is 143\n","[2024-03-08 06:12:04.657]Container exited with a non-zero exit code 143. \n","[2024-03-08 06:12:04.657]Killed by external signal\n",".\n","/usr/lib/spark/python/pyspark/sql/dataframe.py:148: UserWarning: DataFrame.sql_ctx is an internal property, and will be removed in future releases. Use DataFrame.sparkSession instead.\n","  warnings.warn(\n","24/03/08 06:18:18 WARN YarnAllocator: Container from a bad node: container_1709834798656_0014_01_000009 on host: project-ir-selin-dior-w-3.c.project-ir-413718.internal. Exit status: 137. Diagnostics: [2024-03-08 06:18:16.633]Container killed on request. Exit code is 137\n","[2024-03-08 06:18:16.633]Container exited with a non-zero exit code 137. \n","[2024-03-08 06:18:16.633]Killed by external signal\n",".\n","24/03/08 06:18:18 WARN YarnSchedulerBackend$YarnSchedulerEndpoint: Requesting driver to remove executor 8 for reason Container from a bad node: container_1709834798656_0014_01_000009 on host: project-ir-selin-dior-w-3.c.project-ir-413718.internal. Exit status: 137. Diagnostics: [2024-03-08 06:18:16.633]Container killed on request. Exit code is 137\n","[2024-03-08 06:18:16.633]Container exited with a non-zero exit code 137. \n","[2024-03-08 06:18:16.633]Killed by external signal\n",".\n","24/03/08 06:18:18 ERROR YarnScheduler: Lost executor 8 on project-ir-selin-dior-w-3.c.project-ir-413718.internal: Container from a bad node: container_1709834798656_0014_01_000009 on host: project-ir-selin-dior-w-3.c.project-ir-413718.internal. Exit status: 137. Diagnostics: [2024-03-08 06:18:16.633]Container killed on request. Exit code is 137\n","[2024-03-08 06:18:16.633]Container exited with a non-zero exit code 137. \n","[2024-03-08 06:18:16.633]Killed by external signal\n",".\n","24/03/08 06:18:18 WARN TaskSetManager: Lost task 1.0 in stage 22.0 (TID 815) (project-ir-selin-dior-w-3.c.project-ir-413718.internal executor 8): ExecutorLostFailure (executor 8 exited caused by one of the running tasks) Reason: Container from a bad node: container_1709834798656_0014_01_000009 on host: project-ir-selin-dior-w-3.c.project-ir-413718.internal. Exit status: 137. Diagnostics: [2024-03-08 06:18:16.633]Container killed on request. Exit code is 137\n","[2024-03-08 06:18:16.633]Container exited with a non-zero exit code 137. \n","[2024-03-08 06:18:16.633]Killed by external signal\n",".\n","24/03/08 06:19:41 WARN YarnAllocator: Container from a bad node: container_1709834798656_0014_01_000011 on host: project-ir-selin-dior-w-3.c.project-ir-413718.internal. Exit status: 137. Diagnostics: [2024-03-08 06:19:41.867]Container killed on request. Exit code is 137\n","[2024-03-08 06:19:41.867]Container exited with a non-zero exit code 137. \n","[2024-03-08 06:19:41.867]Killed by external signal\n",".\n","24/03/08 06:19:41 ERROR YarnScheduler: Lost executor 10 on project-ir-selin-dior-w-3.c.project-ir-413718.internal: Container from a bad node: container_1709834798656_0014_01_000011 on host: project-ir-selin-dior-w-3.c.project-ir-413718.internal. Exit status: 137. Diagnostics: [2024-03-08 06:19:41.867]Container killed on request. Exit code is 137\n","[2024-03-08 06:19:41.867]Container exited with a non-zero exit code 137. \n","[2024-03-08 06:19:41.867]Killed by external signal\n",".\n","24/03/08 06:19:41 WARN YarnSchedulerBackend$YarnSchedulerEndpoint: Requesting driver to remove executor 10 for reason Container from a bad node: container_1709834798656_0014_01_000011 on host: project-ir-selin-dior-w-3.c.project-ir-413718.internal. Exit status: 137. Diagnostics: [2024-03-08 06:19:41.867]Container killed on request. Exit code is 137\n","[2024-03-08 06:19:41.867]Container exited with a non-zero exit code 137. \n","[2024-03-08 06:19:41.867]Killed by external signal\n",".\n","24/03/08 06:19:41 WARN TaskSetManager: Lost task 0.0 in stage 22.0 (TID 814) (project-ir-selin-dior-w-3.c.project-ir-413718.internal executor 10): ExecutorLostFailure (executor 10 exited caused by one of the running tasks) Reason: Container from a bad node: container_1709834798656_0014_01_000011 on host: project-ir-selin-dior-w-3.c.project-ir-413718.internal. Exit status: 137. Diagnostics: [2024-03-08 06:19:41.867]Container killed on request. Exit code is 137\n","[2024-03-08 06:19:41.867]Container exited with a non-zero exit code 137. \n","[2024-03-08 06:19:41.867]Killed by external signal\n",".\n","24/03/08 06:20:39 WARN YarnAllocator: Container from a bad node: container_1709834798656_0014_01_000012 on host: project-ir-selin-dior-w-3.c.project-ir-413718.internal. Exit status: 137. Diagnostics: [2024-03-08 06:20:39.585]Container killed on request. Exit code is 137\n","[2024-03-08 06:20:39.585]Container exited with a non-zero exit code 137. \n","[2024-03-08 06:20:39.585]Killed by external signal\n",".\n","24/03/08 06:20:39 ERROR YarnScheduler: Lost executor 11 on project-ir-selin-dior-w-3.c.project-ir-413718.internal: Container from a bad node: container_1709834798656_0014_01_000012 on host: project-ir-selin-dior-w-3.c.project-ir-413718.internal. Exit status: 137. Diagnostics: [2024-03-08 06:20:39.585]Container killed on request. Exit code is 137\n","[2024-03-08 06:20:39.585]Container exited with a non-zero exit code 137. \n","[2024-03-08 06:20:39.585]Killed by external signal\n",".\n","24/03/08 06:20:39 WARN TaskSetManager: Lost task 1.1 in stage 22.0 (TID 820) (project-ir-selin-dior-w-3.c.project-ir-413718.internal executor 11): ExecutorLostFailure (executor 11 exited caused by one of the running tasks) Reason: Container from a bad node: container_1709834798656_0014_01_000012 on host: project-ir-selin-dior-w-3.c.project-ir-413718.internal. Exit status: 137. Diagnostics: [2024-03-08 06:20:39.585]Container killed on request. Exit code is 137\n","[2024-03-08 06:20:39.585]Container exited with a non-zero exit code 137. \n","[2024-03-08 06:20:39.585]Killed by external signal\n",".\n","24/03/08 06:20:39 WARN YarnSchedulerBackend$YarnSchedulerEndpoint: Requesting driver to remove executor 11 for reason Container from a bad node: container_1709834798656_0014_01_000012 on host: project-ir-selin-dior-w-3.c.project-ir-413718.internal. Exit status: 137. Diagnostics: [2024-03-08 06:20:39.585]Container killed on request. Exit code is 137\n","[2024-03-08 06:20:39.585]Container exited with a non-zero exit code 137. \n","[2024-03-08 06:20:39.585]Killed by external signal\n",".\n","24/03/08 06:21:31 WARN YarnAllocator: Container from a bad node: container_1709834798656_0014_01_000002 on host: project-ir-selin-dior-w-2.c.project-ir-413718.internal. Exit status: 143. Diagnostics: [2024-03-08 06:21:30.791]Container killed on request. Exit code is 143\n","[2024-03-08 06:21:30.791]Container exited with a non-zero exit code 143. \n","[2024-03-08 06:21:30.792]Killed by external signal\n",".\n","24/03/08 06:21:31 WARN YarnSchedulerBackend$YarnSchedulerEndpoint: Requesting driver to remove executor 2 for reason Container from a bad node: container_1709834798656_0014_01_000002 on host: project-ir-selin-dior-w-2.c.project-ir-413718.internal. Exit status: 143. Diagnostics: [2024-03-08 06:21:30.791]Container killed on request. Exit code is 143\n","[2024-03-08 06:21:30.791]Container exited with a non-zero exit code 143. \n","[2024-03-08 06:21:30.792]Killed by external signal\n",".\n","24/03/08 06:21:31 ERROR YarnScheduler: Lost executor 2 on project-ir-selin-dior-w-2.c.project-ir-413718.internal: Container from a bad node: container_1709834798656_0014_01_000002 on host: project-ir-selin-dior-w-2.c.project-ir-413718.internal. Exit status: 143. Diagnostics: [2024-03-08 06:21:30.791]Container killed on request. Exit code is 143\n","[2024-03-08 06:21:30.791]Container exited with a non-zero exit code 143. \n","[2024-03-08 06:21:30.792]Killed by external signal\n",".\n","24/03/08 06:21:31 WARN TaskSetManager: Lost task 2.0 in stage 21.0 (TID 813) (project-ir-selin-dior-w-2.c.project-ir-413718.internal executor 2): ExecutorLostFailure (executor 2 exited caused by one of the running tasks) Reason: Container from a bad node: container_1709834798656_0014_01_000002 on host: project-ir-selin-dior-w-2.c.project-ir-413718.internal. Exit status: 143. Diagnostics: [2024-03-08 06:21:30.791]Container killed on request. Exit code is 143\n","[2024-03-08 06:21:30.791]Container exited with a non-zero exit code 143. \n","[2024-03-08 06:21:30.792]Killed by external signal\n",".\n","24/03/08 06:21:37 WARN YarnAllocator: Container from a bad node: container_1709834798656_0014_01_000013 on host: project-ir-selin-dior-w-3.c.project-ir-413718.internal. Exit status: 137. Diagnostics: [2024-03-08 06:21:37.599]Container killed on request. Exit code is 137\n","[2024-03-08 06:21:37.599]Container exited with a non-zero exit code 137. \n","[2024-03-08 06:21:37.599]Killed by external signal\n",".\n","24/03/08 06:21:37 WARN YarnSchedulerBackend$YarnSchedulerEndpoint: Requesting driver to remove executor 12 for reason Container from a bad node: container_1709834798656_0014_01_000013 on host: project-ir-selin-dior-w-3.c.project-ir-413718.internal. Exit status: 137. Diagnostics: [2024-03-08 06:21:37.599]Container killed on request. Exit code is 137\n","[2024-03-08 06:21:37.599]Container exited with a non-zero exit code 137. \n","[2024-03-08 06:21:37.599]Killed by external signal\n",".\n","24/03/08 06:21:37 ERROR YarnScheduler: Lost executor 12 on project-ir-selin-dior-w-3.c.project-ir-413718.internal: Container from a bad node: container_1709834798656_0014_01_000013 on host: project-ir-selin-dior-w-3.c.project-ir-413718.internal. Exit status: 137. Diagnostics: [2024-03-08 06:21:37.599]Container killed on request. Exit code is 137\n","[2024-03-08 06:21:37.599]Container exited with a non-zero exit code 137. \n","[2024-03-08 06:21:37.599]Killed by external signal\n",".\n","24/03/08 06:21:37 WARN TaskSetManager: Lost task 0.1 in stage 22.0 (TID 821) (project-ir-selin-dior-w-3.c.project-ir-413718.internal executor 12): ExecutorLostFailure (executor 12 exited caused by one of the running tasks) Reason: Container from a bad node: container_1709834798656_0014_01_000013 on host: project-ir-selin-dior-w-3.c.project-ir-413718.internal. Exit status: 137. Diagnostics: [2024-03-08 06:21:37.599]Container killed on request. Exit code is 137\n","[2024-03-08 06:21:37.599]Container exited with a non-zero exit code 137. \n","[2024-03-08 06:21:37.599]Killed by external signal\n",".\n","24/03/08 06:22:00 WARN HeartbeatReceiver: Removing executor 4 with no recent heartbeats: 165678 ms exceeds timeout 120000 ms\n","24/03/08 06:22:00 ERROR YarnScheduler: Lost executor 4 on project-ir-selin-dior-w-1.c.project-ir-413718.internal: Executor heartbeat timed out after 165678 ms\n","24/03/08 06:22:00 WARN TaskSetManager: Lost task 0.0 in stage 21.0 (TID 811) (project-ir-selin-dior-w-1.c.project-ir-413718.internal executor 4): ExecutorLostFailure (executor 4 exited caused by one of the running tasks) Reason: Executor heartbeat timed out after 165678 ms\n","24/03/08 06:22:02 WARN TransportChannelHandler: Exception in connection from /10.128.0.58:53666\n","java.io.IOException: Connection reset by peer\n","\tat sun.nio.ch.FileDispatcherImpl.read0(Native Method) ~[?:?]\n","\tat sun.nio.ch.SocketDispatcher.read(SocketDispatcher.java:39) ~[?:?]\n","\tat sun.nio.ch.IOUtil.readIntoNativeBuffer(IOUtil.java:276) ~[?:?]\n","\tat sun.nio.ch.IOUtil.read(IOUtil.java:233) ~[?:?]\n","\tat sun.nio.ch.IOUtil.read(IOUtil.java:223) ~[?:?]\n","\tat sun.nio.ch.SocketChannelImpl.read(SocketChannelImpl.java:356) ~[?:?]\n","\tat io.netty.buffer.PooledByteBuf.setBytes(PooledByteBuf.java:258) ~[netty-buffer-4.1.77.Final.jar:4.1.77.Final]\n","\tat io.netty.buffer.AbstractByteBuf.writeBytes(AbstractByteBuf.java:1132) ~[netty-buffer-4.1.77.Final.jar:4.1.77.Final]\n","\tat io.netty.channel.socket.nio.NioSocketChannel.doReadBytes(NioSocketChannel.java:357) ~[netty-transport-4.1.77.Final.jar:4.1.77.Final]\n","\tat io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:151) ~[netty-transport-4.1.77.Final.jar:4.1.77.Final]\n","\tat io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:722) ~[netty-transport-4.1.77.Final.jar:4.1.77.Final]\n","\tat io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:658) ~[netty-transport-4.1.77.Final.jar:4.1.77.Final]\n","\tat io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:584) ~[netty-transport-4.1.77.Final.jar:4.1.77.Final]\n","\tat io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:496) ~[netty-transport-4.1.77.Final.jar:4.1.77.Final]\n","\tat io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:995) ~[netty-common-4.1.77.Final.jar:4.1.77.Final]\n","\tat io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) ~[netty-common-4.1.77.Final.jar:4.1.77.Final]\n","\tat io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) ~[netty-common-4.1.77.Final.jar:4.1.77.Final]\n","\tat java.lang.Thread.run(Thread.java:829) ~[?:?]\n","24/03/08 06:22:13 WARN YarnAllocator: Container from a bad node: container_1709834798656_0014_01_000008 on host: project-ir-selin-dior-w-0.c.project-ir-413718.internal. Exit status: 143. Diagnostics: [2024-03-08 06:22:13.233]Container killed on request. Exit code is 143\n","[2024-03-08 06:22:13.233]Container exited with a non-zero exit code 143. \n","[2024-03-08 06:22:13.234]Killed by external signal\n",".\n","24/03/08 06:22:13 WARN YarnSchedulerBackend$YarnSchedulerEndpoint: Requesting driver to remove executor 7 for reason Container from a bad node: container_1709834798656_0014_01_000008 on host: project-ir-selin-dior-w-0.c.project-ir-413718.internal. Exit status: 143. Diagnostics: [2024-03-08 06:22:13.233]Container killed on request. Exit code is 143\n","[2024-03-08 06:22:13.233]Container exited with a non-zero exit code 143. \n","[2024-03-08 06:22:13.234]Killed by external signal\n",".\n","24/03/08 06:22:13 ERROR YarnScheduler: Lost executor 7 on project-ir-selin-dior-w-0.c.project-ir-413718.internal: Container from a bad node: container_1709834798656_0014_01_000008 on host: project-ir-selin-dior-w-0.c.project-ir-413718.internal. Exit status: 143. Diagnostics: [2024-03-08 06:22:13.233]Container killed on request. Exit code is 143\n","[2024-03-08 06:22:13.233]Container exited with a non-zero exit code 143. \n","[2024-03-08 06:22:13.234]Killed by external signal\n",".\n","24/03/08 06:22:13 WARN TaskSetManager: Lost task 1.0 in stage 21.0 (TID 812) (project-ir-selin-dior-w-0.c.project-ir-413718.internal executor 7): ExecutorLostFailure (executor 7 exited caused by one of the running tasks) Reason: Container from a bad node: container_1709834798656_0014_01_000008 on host: project-ir-selin-dior-w-0.c.project-ir-413718.internal. Exit status: 143. Diagnostics: [2024-03-08 06:22:13.233]Container killed on request. Exit code is 143\n","[2024-03-08 06:22:13.233]Container exited with a non-zero exit code 143. \n","[2024-03-08 06:22:13.234]Killed by external signal\n",".\n","24/03/08 06:22:32 WARN YarnAllocator: Container from a bad node: container_1709834798656_0014_01_000014 on host: project-ir-selin-dior-w-3.c.project-ir-413718.internal. Exit status: 137. Diagnostics: [2024-03-08 06:22:32.566]Container killed on request. Exit code is 137\n","[2024-03-08 06:22:32.566]Container exited with a non-zero exit code 137. \n","[2024-03-08 06:22:32.567]Killed by external signal\n",".\n","24/03/08 06:22:32 WARN YarnSchedulerBackend$YarnSchedulerEndpoint: Requesting driver to remove executor 13 for reason Container from a bad node: container_1709834798656_0014_01_000014 on host: project-ir-selin-dior-w-3.c.project-ir-413718.internal. Exit status: 137. Diagnostics: [2024-03-08 06:22:32.566]Container killed on request. Exit code is 137\n","[2024-03-08 06:22:32.566]Container exited with a non-zero exit code 137. \n","[2024-03-08 06:22:32.567]Killed by external signal\n",".\n","24/03/08 06:22:32 ERROR YarnScheduler: Lost executor 13 on project-ir-selin-dior-w-3.c.project-ir-413718.internal: Container from a bad node: container_1709834798656_0014_01_000014 on host: project-ir-selin-dior-w-3.c.project-ir-413718.internal. Exit status: 137. Diagnostics: [2024-03-08 06:22:32.566]Container killed on request. Exit code is 137\n","[2024-03-08 06:22:32.566]Container exited with a non-zero exit code 137. \n","[2024-03-08 06:22:32.567]Killed by external signal\n",".\n","24/03/08 06:22:32 WARN TaskSetManager: Lost task 1.2 in stage 22.0 (TID 822) (project-ir-selin-dior-w-3.c.project-ir-413718.internal executor 13): ExecutorLostFailure (executor 13 exited caused by one of the running tasks) Reason: Container from a bad node: container_1709834798656_0014_01_000014 on host: project-ir-selin-dior-w-3.c.project-ir-413718.internal. Exit status: 137. Diagnostics: [2024-03-08 06:22:32.566]Container killed on request. Exit code is 137\n","[2024-03-08 06:22:32.566]Container exited with a non-zero exit code 137. \n","[2024-03-08 06:22:32.567]Killed by external signal\n",".\n","24/03/08 06:23:30 WARN YarnAllocator: Container from a bad node: container_1709834798656_0014_01_000016 on host: project-ir-selin-dior-w-3.c.project-ir-413718.internal. Exit status: 137. Diagnostics: [2024-03-08 06:23:30.227]Container killed on request. Exit code is 137\n","[2024-03-08 06:23:30.227]Container exited with a non-zero exit code 137. \n","[2024-03-08 06:23:30.228]Killed by external signal\n",".\n","24/03/08 06:23:30 WARN YarnSchedulerBackend$YarnSchedulerEndpoint: Requesting driver to remove executor 15 for reason Container from a bad node: container_1709834798656_0014_01_000016 on host: project-ir-selin-dior-w-3.c.project-ir-413718.internal. Exit status: 137. Diagnostics: [2024-03-08 06:23:30.227]Container killed on request. Exit code is 137\n","[2024-03-08 06:23:30.227]Container exited with a non-zero exit code 137. \n","[2024-03-08 06:23:30.228]Killed by external signal\n",".\n","24/03/08 06:23:30 ERROR YarnScheduler: Lost executor 15 on project-ir-selin-dior-w-3.c.project-ir-413718.internal: Container from a bad node: container_1709834798656_0014_01_000016 on host: project-ir-selin-dior-w-3.c.project-ir-413718.internal. Exit status: 137. Diagnostics: [2024-03-08 06:23:30.227]Container killed on request. Exit code is 137\n","[2024-03-08 06:23:30.227]Container exited with a non-zero exit code 137. \n","[2024-03-08 06:23:30.228]Killed by external signal\n",".\n","24/03/08 06:23:30 WARN TaskSetManager: Lost task 0.2 in stage 22.0 (TID 824) (project-ir-selin-dior-w-3.c.project-ir-413718.internal executor 15): ExecutorLostFailure (executor 15 exited caused by one of the running tasks) Reason: Container from a bad node: container_1709834798656_0014_01_000016 on host: project-ir-selin-dior-w-3.c.project-ir-413718.internal. Exit status: 137. Diagnostics: [2024-03-08 06:23:30.227]Container killed on request. Exit code is 137\n","[2024-03-08 06:23:30.227]Container exited with a non-zero exit code 137. \n","[2024-03-08 06:23:30.228]Killed by external signal\n",".\n","24/03/08 06:24:27 WARN YarnAllocator: Container from a bad node: container_1709834798656_0014_01_000019 on host: project-ir-selin-dior-w-3.c.project-ir-413718.internal. Exit status: 137. Diagnostics: [2024-03-08 06:24:27.912]Container killed on request. Exit code is 137\n","[2024-03-08 06:24:27.912]Container exited with a non-zero exit code 137. \n","[2024-03-08 06:24:27.912]Killed by external signal\n",".\n","24/03/08 06:24:27 WARN YarnSchedulerBackend$YarnSchedulerEndpoint: Requesting driver to remove executor 18 for reason Container from a bad node: container_1709834798656_0014_01_000019 on host: project-ir-selin-dior-w-3.c.project-ir-413718.internal. Exit status: 137. Diagnostics: [2024-03-08 06:24:27.912]Container killed on request. Exit code is 137\n","[2024-03-08 06:24:27.912]Container exited with a non-zero exit code 137. \n","[2024-03-08 06:24:27.912]Killed by external signal\n",".\n","24/03/08 06:24:27 ERROR YarnScheduler: Lost executor 18 on project-ir-selin-dior-w-3.c.project-ir-413718.internal: Container from a bad node: container_1709834798656_0014_01_000019 on host: project-ir-selin-dior-w-3.c.project-ir-413718.internal. Exit status: 137. Diagnostics: [2024-03-08 06:24:27.912]Container killed on request. Exit code is 137\n","[2024-03-08 06:24:27.912]Container exited with a non-zero exit code 137. \n","[2024-03-08 06:24:27.912]Killed by external signal\n",".\n","24/03/08 06:24:27 WARN TaskSetManager: Lost task 1.3 in stage 22.0 (TID 827) (project-ir-selin-dior-w-3.c.project-ir-413718.internal executor 18): ExecutorLostFailure (executor 18 exited caused by one of the running tasks) Reason: Container from a bad node: container_1709834798656_0014_01_000019 on host: project-ir-selin-dior-w-3.c.project-ir-413718.internal. Exit status: 137. Diagnostics: [2024-03-08 06:24:27.912]Container killed on request. Exit code is 137\n","[2024-03-08 06:24:27.912]Container exited with a non-zero exit code 137. \n","[2024-03-08 06:24:27.912]Killed by external signal\n",".\n","24/03/08 06:24:27 ERROR TaskSetManager: Task 1 in stage 22.0 failed 4 times; aborting job\n","[Stage 21:>                 (0 + 3) / 4][Stage 22:>                 (0 + 1) / 4]\r"]},{"ename":"Py4JJavaError","evalue":"An error occurred while calling o203.run.\n: org.apache.spark.SparkException: Job aborted due to stage failure: Task 1 in stage 22.0 failed 4 times, most recent failure: Lost task 1.3 in stage 22.0 (TID 827) (project-ir-selin-dior-w-3.c.project-ir-413718.internal executor 18): ExecutorLostFailure (executor 18 exited caused by one of the running tasks) Reason: Container from a bad node: container_1709834798656_0014_01_000019 on host: project-ir-selin-dior-w-3.c.project-ir-413718.internal. Exit status: 137. Diagnostics: [2024-03-08 06:24:27.912]Container killed on request. Exit code is 137\n[2024-03-08 06:24:27.912]Container exited with a non-zero exit code 137. \n[2024-03-08 06:24:27.912]Killed by external signal\n.\nDriver stacktrace:\n\tat org.apache.spark.scheduler.DAGScheduler.failJobAndIndependentStages(DAGScheduler.scala:2717)\n\tat org.apache.spark.scheduler.DAGScheduler.$anonfun$abortStage$2(DAGScheduler.scala:2653)\n\tat org.apache.spark.scheduler.DAGScheduler.$anonfun$abortStage$2$adapted(DAGScheduler.scala:2652)\n\tat scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:62)\n\tat scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:55)\n\tat scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:49)\n\tat org.apache.spark.scheduler.DAGScheduler.abortStage(DAGScheduler.scala:2652)\n\tat org.apache.spark.scheduler.DAGScheduler.$anonfun$handleTaskSetFailed$1(DAGScheduler.scala:1189)\n\tat org.apache.spark.scheduler.DAGScheduler.$anonfun$handleTaskSetFailed$1$adapted(DAGScheduler.scala:1189)\n\tat scala.Option.foreach(Option.scala:407)\n\tat org.apache.spark.scheduler.DAGScheduler.handleTaskSetFailed(DAGScheduler.scala:1189)\n\tat org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.doOnReceive(DAGScheduler.scala:2913)\n\tat org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2855)\n\tat org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2844)\n\tat org.apache.spark.util.EventLoop$$anon$1.run(EventLoop.scala:49)\n\tat org.apache.spark.scheduler.DAGScheduler.runJob(DAGScheduler.scala:959)\n\tat org.apache.spark.SparkContext.runJob(SparkContext.scala:2293)\n\tat org.apache.spark.SparkContext.runJob(SparkContext.scala:2314)\n\tat org.apache.spark.SparkContext.runJob(SparkContext.scala:2333)\n\tat org.apache.spark.SparkContext.runJob(SparkContext.scala:2358)\n\tat org.apache.spark.rdd.RDD.$anonfun$foreachPartition$1(RDD.scala:1011)\n\tat org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)\n\tat org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:112)\n\tat org.apache.spark.rdd.RDD.withScope(RDD.scala:406)\n\tat org.apache.spark.rdd.RDD.foreachPartition(RDD.scala:1009)\n\tat org.apache.spark.graphx.lib.PageRank$.runWithOptions(PageRank.scala:199)\n\tat org.apache.spark.graphx.lib.PageRank$.runWithOptions(PageRank.scala:144)\n\tat org.graphframes.lib.PageRank$.run(PageRank.scala:130)\n\tat org.graphframes.lib.PageRank.run(PageRank.scala:104)\n\tat java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)\n\tat java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)\n\tat java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)\n\tat java.base/java.lang.reflect.Method.invoke(Method.java:566)\n\tat py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)\n\tat py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:357)\n\tat py4j.Gateway.invoke(Gateway.java:282)\n\tat py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)\n\tat py4j.commands.CallCommand.execute(CallCommand.java:79)\n\tat py4j.ClientServerConnection.waitForCommands(ClientServerConnection.java:182)\n\tat py4j.ClientServerConnection.run(ClientServerConnection.java:106)\n\tat java.base/java.lang.Thread.run(Thread.java:829)\n","output_type":"error","traceback":["\u001B[0;31m---------------------------------------------------------------------------\u001B[0m","\u001B[0;31mPy4JJavaError\u001B[0m                             Traceback (most recent call last)","\u001B[0;32m/tmp/ipykernel_105374/3486764191.py\u001B[0m in \u001B[0;36m<cell line: 10>\u001B[0;34m()\u001B[0m\n\u001B[1;32m      8\u001B[0m \u001B[0mverticesDF\u001B[0m \u001B[0;34m=\u001B[0m \u001B[0mvertices\u001B[0m\u001B[0;34m.\u001B[0m\u001B[0mtoDF\u001B[0m\u001B[0;34m(\u001B[0m\u001B[0;34m[\u001B[0m\u001B[0;34m'id'\u001B[0m\u001B[0;34m]\u001B[0m\u001B[0;34m)\u001B[0m\u001B[0;34m.\u001B[0m\u001B[0mrepartition\u001B[0m\u001B[0;34m(\u001B[0m\u001B[0;36m4\u001B[0m\u001B[0;34m,\u001B[0m \u001B[0;34m'id'\u001B[0m\u001B[0;34m)\u001B[0m\u001B[0;34m\u001B[0m\u001B[0;34m\u001B[0m\u001B[0m\n\u001B[1;32m      9\u001B[0m \u001B[0mg\u001B[0m \u001B[0;34m=\u001B[0m \u001B[0mGraphFrame\u001B[0m\u001B[0;34m(\u001B[0m\u001B[0mverticesDF\u001B[0m\u001B[0;34m,\u001B[0m \u001B[0medgesDF\u001B[0m\u001B[0;34m)\u001B[0m\u001B[0;34m\u001B[0m\u001B[0;34m\u001B[0m\u001B[0m\n\u001B[0;32m---> 10\u001B[0;31m \u001B[0mpr_results\u001B[0m \u001B[0;34m=\u001B[0m \u001B[0mg\u001B[0m\u001B[0;34m.\u001B[0m\u001B[0mpageRank\u001B[0m\u001B[0;34m(\u001B[0m\u001B[0mresetProbability\u001B[0m\u001B[0;34m=\u001B[0m\u001B[0;36m0.15\u001B[0m\u001B[0;34m,\u001B[0m \u001B[0mmaxIter\u001B[0m\u001B[0;34m=\u001B[0m\u001B[0;36m10\u001B[0m\u001B[0;34m)\u001B[0m\u001B[0;34m\u001B[0m\u001B[0;34m\u001B[0m\u001B[0m\n\u001B[0m","\u001B[0;32m/opt/conda/miniconda3/lib/python3.10/site-packages/graphframes/graphframe.py\u001B[0m in \u001B[0;36mpageRank\u001B[0;34m(self, resetProbability, sourceId, maxIter, tol)\u001B[0m\n\u001B[1;32m    357\u001B[0m             \u001B[0;32massert\u001B[0m \u001B[0mtol\u001B[0m \u001B[0;32mis\u001B[0m \u001B[0;32mnot\u001B[0m \u001B[0;32mNone\u001B[0m\u001B[0;34m,\u001B[0m \u001B[0;34m\"Exactly one of maxIter or tol should be set.\"\u001B[0m\u001B[0;34m\u001B[0m\u001B[0;34m\u001B[0m\u001B[0m\n\u001B[1;32m    358\u001B[0m             \u001B[0mbuilder\u001B[0m \u001B[0;34m=\u001B[0m \u001B[0mbuilder\u001B[0m\u001B[0;34m.\u001B[0m\u001B[0mtol\u001B[0m\u001B[0;34m(\u001B[0m\u001B[0mtol\u001B[0m\u001B[0;34m)\u001B[0m\u001B[0;34m\u001B[0m\u001B[0;34m\u001B[0m\u001B[0m\n\u001B[0;32m--> 359\u001B[0;31m         \u001B[0mjgf\u001B[0m \u001B[0;34m=\u001B[0m \u001B[0mbuilder\u001B[0m\u001B[0;34m.\u001B[0m\u001B[0mrun\u001B[0m\u001B[0;34m(\u001B[0m\u001B[0;34m)\u001B[0m\u001B[0;34m\u001B[0m\u001B[0;34m\u001B[0m\u001B[0m\n\u001B[0m\u001B[1;32m    360\u001B[0m         \u001B[0;32mreturn\u001B[0m \u001B[0m_from_java_gf\u001B[0m\u001B[0;34m(\u001B[0m\u001B[0mjgf\u001B[0m\u001B[0;34m,\u001B[0m \u001B[0mself\u001B[0m\u001B[0;34m.\u001B[0m\u001B[0m_sqlContext\u001B[0m\u001B[0;34m)\u001B[0m\u001B[0;34m\u001B[0m\u001B[0;34m\u001B[0m\u001B[0m\n\u001B[1;32m    361\u001B[0m \u001B[0;34m\u001B[0m\u001B[0m\n","\u001B[0;32m/usr/lib/spark/python/lib/py4j-0.10.9.5-src.zip/py4j/java_gateway.py\u001B[0m in \u001B[0;36m__call__\u001B[0;34m(self, *args)\u001B[0m\n\u001B[1;32m   1319\u001B[0m \u001B[0;34m\u001B[0m\u001B[0m\n\u001B[1;32m   1320\u001B[0m         \u001B[0manswer\u001B[0m \u001B[0;34m=\u001B[0m \u001B[0mself\u001B[0m\u001B[0;34m.\u001B[0m\u001B[0mgateway_client\u001B[0m\u001B[0;34m.\u001B[0m\u001B[0msend_command\u001B[0m\u001B[0;34m(\u001B[0m\u001B[0mcommand\u001B[0m\u001B[0;34m)\u001B[0m\u001B[0;34m\u001B[0m\u001B[0;34m\u001B[0m\u001B[0m\n\u001B[0;32m-> 1321\u001B[0;31m         return_value = get_return_value(\n\u001B[0m\u001B[1;32m   1322\u001B[0m             answer, self.gateway_client, self.target_id, self.name)\n\u001B[1;32m   1323\u001B[0m \u001B[0;34m\u001B[0m\u001B[0m\n","\u001B[0;32m/usr/lib/spark/python/pyspark/sql/utils.py\u001B[0m in \u001B[0;36mdeco\u001B[0;34m(*a, **kw)\u001B[0m\n\u001B[1;32m    188\u001B[0m     \u001B[0;32mdef\u001B[0m \u001B[0mdeco\u001B[0m\u001B[0;34m(\u001B[0m\u001B[0;34m*\u001B[0m\u001B[0ma\u001B[0m\u001B[0;34m:\u001B[0m \u001B[0mAny\u001B[0m\u001B[0;34m,\u001B[0m \u001B[0;34m**\u001B[0m\u001B[0mkw\u001B[0m\u001B[0;34m:\u001B[0m \u001B[0mAny\u001B[0m\u001B[0;34m)\u001B[0m \u001B[0;34m->\u001B[0m \u001B[0mAny\u001B[0m\u001B[0;34m:\u001B[0m\u001B[0;34m\u001B[0m\u001B[0;34m\u001B[0m\u001B[0m\n\u001B[1;32m    189\u001B[0m         \u001B[0;32mtry\u001B[0m\u001B[0;34m:\u001B[0m\u001B[0;34m\u001B[0m\u001B[0;34m\u001B[0m\u001B[0m\n\u001B[0;32m--> 190\u001B[0;31m             \u001B[0;32mreturn\u001B[0m \u001B[0mf\u001B[0m\u001B[0;34m(\u001B[0m\u001B[0;34m*\u001B[0m\u001B[0ma\u001B[0m\u001B[0;34m,\u001B[0m \u001B[0;34m**\u001B[0m\u001B[0mkw\u001B[0m\u001B[0;34m)\u001B[0m\u001B[0;34m\u001B[0m\u001B[0;34m\u001B[0m\u001B[0m\n\u001B[0m\u001B[1;32m    191\u001B[0m         \u001B[0;32mexcept\u001B[0m \u001B[0mPy4JJavaError\u001B[0m \u001B[0;32mas\u001B[0m \u001B[0me\u001B[0m\u001B[0;34m:\u001B[0m\u001B[0;34m\u001B[0m\u001B[0;34m\u001B[0m\u001B[0m\n\u001B[1;32m    192\u001B[0m             \u001B[0mconverted\u001B[0m \u001B[0;34m=\u001B[0m \u001B[0mconvert_exception\u001B[0m\u001B[0;34m(\u001B[0m\u001B[0me\u001B[0m\u001B[0;34m.\u001B[0m\u001B[0mjava_exception\u001B[0m\u001B[0;34m)\u001B[0m\u001B[0;34m\u001B[0m\u001B[0;34m\u001B[0m\u001B[0m\n","\u001B[0;32m/usr/lib/spark/python/lib/py4j-0.10.9.5-src.zip/py4j/protocol.py\u001B[0m in \u001B[0;36mget_return_value\u001B[0;34m(answer, gateway_client, target_id, name)\u001B[0m\n\u001B[1;32m    324\u001B[0m             \u001B[0mvalue\u001B[0m \u001B[0;34m=\u001B[0m \u001B[0mOUTPUT_CONVERTER\u001B[0m\u001B[0;34m[\u001B[0m\u001B[0mtype\u001B[0m\u001B[0;34m]\u001B[0m\u001B[0;34m(\u001B[0m\u001B[0manswer\u001B[0m\u001B[0;34m[\u001B[0m\u001B[0;36m2\u001B[0m\u001B[0;34m:\u001B[0m\u001B[0;34m]\u001B[0m\u001B[0;34m,\u001B[0m \u001B[0mgateway_client\u001B[0m\u001B[0;34m)\u001B[0m\u001B[0;34m\u001B[0m\u001B[0;34m\u001B[0m\u001B[0m\n\u001B[1;32m    325\u001B[0m             \u001B[0;32mif\u001B[0m \u001B[0manswer\u001B[0m\u001B[0;34m[\u001B[0m\u001B[0;36m1\u001B[0m\u001B[0;34m]\u001B[0m \u001B[0;34m==\u001B[0m \u001B[0mREFERENCE_TYPE\u001B[0m\u001B[0;34m:\u001B[0m\u001B[0;34m\u001B[0m\u001B[0;34m\u001B[0m\u001B[0m\n\u001B[0;32m--> 326\u001B[0;31m                 raise Py4JJavaError(\n\u001B[0m\u001B[1;32m    327\u001B[0m                     \u001B[0;34m\"An error occurred while calling {0}{1}{2}.\\n\"\u001B[0m\u001B[0;34m.\u001B[0m\u001B[0;34m\u001B[0m\u001B[0;34m\u001B[0m\u001B[0m\n\u001B[1;32m    328\u001B[0m                     format(target_id, \".\", name), value)\n","\u001B[0;31mPy4JJavaError\u001B[0m: An error occurred while calling o203.run.\n: org.apache.spark.SparkException: Job aborted due to stage failure: Task 1 in stage 22.0 failed 4 times, most recent failure: Lost task 1.3 in stage 22.0 (TID 827) (project-ir-selin-dior-w-3.c.project-ir-413718.internal executor 18): ExecutorLostFailure (executor 18 exited caused by one of the running tasks) Reason: Container from a bad node: container_1709834798656_0014_01_000019 on host: project-ir-selin-dior-w-3.c.project-ir-413718.internal. Exit status: 137. Diagnostics: [2024-03-08 06:24:27.912]Container killed on request. Exit code is 137\n[2024-03-08 06:24:27.912]Container exited with a non-zero exit code 137. \n[2024-03-08 06:24:27.912]Killed by external signal\n.\nDriver stacktrace:\n\tat org.apache.spark.scheduler.DAGScheduler.failJobAndIndependentStages(DAGScheduler.scala:2717)\n\tat org.apache.spark.scheduler.DAGScheduler.$anonfun$abortStage$2(DAGScheduler.scala:2653)\n\tat org.apache.spark.scheduler.DAGScheduler.$anonfun$abortStage$2$adapted(DAGScheduler.scala:2652)\n\tat scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:62)\n\tat scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:55)\n\tat scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:49)\n\tat org.apache.spark.scheduler.DAGScheduler.abortStage(DAGScheduler.scala:2652)\n\tat org.apache.spark.scheduler.DAGScheduler.$anonfun$handleTaskSetFailed$1(DAGScheduler.scala:1189)\n\tat org.apache.spark.scheduler.DAGScheduler.$anonfun$handleTaskSetFailed$1$adapted(DAGScheduler.scala:1189)\n\tat scala.Option.foreach(Option.scala:407)\n\tat org.apache.spark.scheduler.DAGScheduler.handleTaskSetFailed(DAGScheduler.scala:1189)\n\tat org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.doOnReceive(DAGScheduler.scala:2913)\n\tat org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2855)\n\tat org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2844)\n\tat org.apache.spark.util.EventLoop$$anon$1.run(EventLoop.scala:49)\n\tat org.apache.spark.scheduler.DAGScheduler.runJob(DAGScheduler.scala:959)\n\tat org.apache.spark.SparkContext.runJob(SparkContext.scala:2293)\n\tat org.apache.spark.SparkContext.runJob(SparkContext.scala:2314)\n\tat org.apache.spark.SparkContext.runJob(SparkContext.scala:2333)\n\tat org.apache.spark.SparkContext.runJob(SparkContext.scala:2358)\n\tat org.apache.spark.rdd.RDD.$anonfun$foreachPartition$1(RDD.scala:1011)\n\tat org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)\n\tat org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:112)\n\tat org.apache.spark.rdd.RDD.withScope(RDD.scala:406)\n\tat org.apache.spark.rdd.RDD.foreachPartition(RDD.scala:1009)\n\tat org.apache.spark.graphx.lib.PageRank$.runWithOptions(PageRank.scala:199)\n\tat org.apache.spark.graphx.lib.PageRank$.runWithOptions(PageRank.scala:144)\n\tat org.graphframes.lib.PageRank$.run(PageRank.scala:130)\n\tat org.graphframes.lib.PageRank.run(PageRank.scala:104)\n\tat java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)\n\tat java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)\n\tat java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)\n\tat java.base/java.lang.reflect.Method.invoke(Method.java:566)\n\tat py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)\n\tat py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:357)\n\tat py4j.Gateway.invoke(Gateway.java:282)\n\tat py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)\n\tat py4j.commands.CallCommand.execute(CallCommand.java:79)\n\tat py4j.ClientServerConnection.waitForCommands(ClientServerConnection.java:182)\n\tat py4j.ClientServerConnection.run(ClientServerConnection.java:106)\n\tat java.base/java.lang.Thread.run(Thread.java:829)\n"]}],"source":["# construct the graph for a small sample of (1000) pages\n","edges, vertices = generate_graph(pages_links)\n","\n","# time the actual execution\n","v_cnt, e_cnt = vertices.count(), edges.count()\n","\n","edgesDF = edges.toDF(['src', 'dst']).repartition(4, 'src')\n","verticesDF = vertices.toDF(['id']).repartition(4, 'id')\n","g = GraphFrame(verticesDF, edgesDF)\n","pr_results = g.pageRank(resetProbability=0.15, maxIter=10)"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"49gwy16_Cguo"},"outputs":[],"source":["pr = pr_results.vertices.select(\"id\", \"pagerank\")"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["pr_dict = pr.rdd.map(lambda row: (row.id, row.pagerank)).collectAsMap()"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"q35rtBGDCwS6"},"outputs":[],"source":["name = \"pagerank_dict\"\n","with open(f\"{name}.pkl\", \"wb\") as f:\n","    pickle.dump(pr_dict, f)\n","\n","client = storage.Client()\n","bucket = client.bucket(bucket_name)\n","blob_posting_locs = bucket.blob(f\"page_rank/{name}.pkl\")\n","blob_posting_locs.upload_from_filename(f\"{name}.pkl\")"]}],"metadata":{"colab":{"provenance":[]},"kernelspec":{"display_name":"PySpark","language":"python","name":"pyspark"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.10.8"}},"nbformat":4,"nbformat_minor":1}